### PR TITLE
LinkageCheckerMain should not fail upon missing Maven artifact

### LINKED ISSUE #1242: LinkageCheckerMain should not fail upon missing Artifact
```
> Task :checkJavaLinkage
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Exception in thread "main" org.eclipse.aether.RepositoryException: Unresolved artifacts: xerces:xerces-impl:jar:2.6.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > cglib:cglib-nodep:jar:2.2.2 (provided) > ant:ant:jar:1.6.2 (compile?) > xerces:xerces-impl:jar:2.6.2 (compile?), xml-apis:xml-apis:jar:2.6.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > cglib:cglib-nodep:jar:2.2.2 (provided) > ant:ant:jar:1.6.2 (compile?) > xml-apis:xml-apis:jar:2.6.2 (compile?), maven-plugins:maven-cobertura-plugin:plugin:1.3 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework:spring-beans:jar:4.1.6.RELEASE (compile?) > org.codehaus.groovy:groovy-all:jar:2.3.11 (compile?) > com.thoughtworks.xstream:xstream:jar:1.4.7 (compile?) > org.jdom:jdom:jar:1.1.3 (compile?) > jaxen:jaxen:jar:1.1.3 (compile?) > dom4j:dom4j:jar:1.6.1 (compile) > jaxen:jaxen:jar:1.1-beta-6 (compile?) > jdom:jdom:jar:1.0 (compile) > jaxen:jaxen:jar:1.0-FCS (compile?) > maven-plugins:maven-cobertura-plugin:plugin:1.3 (compile), maven-plugins:maven-findbugs-plugin:plugin:1.3.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework:spring-beans:jar:4.1.6.RELEASE (compile?) > org.codehaus.groovy:groovy-all:jar:2.3.11 (compile?) > com.thoughtworks.xstream:xstream:jar:1.4.7 (compile?) > org.jdom:jdom:jar:1.1.3 (compile?) > jaxen:jaxen:jar:1.1.3 (compile?) > dom4j:dom4j:jar:1.6.1 (compile) > jaxen:jaxen:jar:1.1-beta-6 (compile?) > jdom:jdom:jar:1.0 (compile) > jaxen:jaxen:jar:1.0-FCS (compile?) > maven-plugins:maven-findbugs-plugin:plugin:1.3.1 (compile), com.oracle.oc4j:com.springsource.oracle.classloader:jar:10.1.3.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.context:jar:2.5.6.SEC01 (compile) > com.oracle.oc4j:com.springsource.oracle.classloader:jar:10.1.3.1 (provided), com.oracle.jdbc:com.springsource.oracle.jdbc:jar:10.2.0.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.jdbc:jar:2.5.6.SEC01 (compile?) > com.experlog.xapool:com.springsource.org.enhydra.jdbc:jar:1.5.0 (compile?) > com.oracle.jdbc:com.springsource.oracle.jdbc:jar:10.2.0.2 (compile?), com.oracle.toplink:com.springsource.oracle.toplink:jar:10.1.3 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > com.oracle.toplink:com.springsource.oracle.toplink:jar:10.1.3 (compile?), com.oracle.aq:com.springsource.oracle.aq:jar:10.1.0.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.aq:com.springsource.oracle.aq:jar:10.1.0.1 (compile?), com.oracle.spatial:com.springsource.oracle.spatial.geometry:jar:10.2.0.1 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.spatial:com.springsource.oracle.spatial.geometry:jar:10.2.0.1 (compile?), com.oracle.xdb:com.springsource.oracle.xdb:jar:10.2.0.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.xdb:com.springsource.oracle.xdb:jar:10.2.0.2 (compile?), com.oracle.xml:com.springsource.oracle.xml:jar:10.2.0.2 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.apache.cxf:cxf-rt-transports-http:jar:3.1.2 (provided?) > org.apache.cxf:cxf-core:jar:3.1.2 (compile) > org.springframework.osgi:spring-osgi-core:jar:1.2.1 (provided?) > org.springframework:org.springframework.test:jar:2.5.6.SEC01 (provided) > org.springframework:org.springframework.orm:jar:2.5.6.SEC01 (compile?) > org.eclipse.persistence:com.springsource.org.eclipse.persistence:jar:1.0.0.m5 (compile?) > com.oracle.xml:com.springsource.oracle.xml:jar:10.2.0.2 (compile?), com.jcraft:jsch.agentproxy:jar:0.0.6 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-context:jar:4.3.7.RELEASE (compile) > org.codehaus.groovy:groovy-all:jar:2.4.9 (compile?) > org.apache.ivy:ivy:jar:2.4.0 (compile?) > com.jcraft:jsch.agentproxy:jar:0.0.6 (compile?), javax.security:jaas:jar:1.0.01 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-webmvc:jar:4.3.7.RELEASE (compile?) > net.sf.jasperreports:jasperreports:jar:6.2.1 (compile?) > org.hibernate:hibernate-core:jar:3.3.2.GA (compile?) > javax.security:jaas:jar:1.0.01 (provided), javax.security:jacc:jar:1.0 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-webmvc:jar:4.3.7.RELEASE (compile?) > net.sf.jasperreports:jasperreports:jar:6.2.1 (compile?) > org.hibernate:hibernate-core:jar:3.3.2.GA (compile?) > javax.security:jacc:jar:1.0 (provided), com.ibm.websphere:uow:jar:6.0.2.17 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework:spring-websocket:jar:4.3.7.RELEASE (provided?) > org.springframework:spring-webmvc:jar:4.3.7.RELEASE (compile?) > org.springframework:spring-context-support:jar:4.3.7.RELEASE (compile?) > org.springframework:spring-jdbc:jar:4.3.7.RELEASE (compile?) > org.springframework:spring-tx:jar:4.3.7.RELEASE (compile) > com.ibm.websphere:uow:jar:6.0.2.17 (compile?), org.jruby:readline:jar:1.0 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework.data:spring-data-redis:jar:1.8.6.RELEASE (provided?) > org.springframework.data:spring-data-keyvalue:jar:1.2.6.RELEASE (compile) > org.springframework.data:spring-data-commons:jar:1.13.6.RELEASE (compile) > org.springframework.hateoas:spring-hateoas:jar:0.23.0.RELEASE (compile?) > org.springframework:spring-aop:jar:4.3.5.RELEASE (compile) > com.jamonapi:jamon:jar:2.81 (compile?) > org.springframework:spring-context:jar:4.0.2.RELEASE (provided) > org.jruby:jruby:jar:1.7.10 (compile?) > org.jruby:jruby-stdlib:jar:1.7.10 (compile) > org.jruby:readline:jar:1.0 (provided), org.jruby:openssl:jar:0.9.4 was not resolved. Dependency path: org.apache.beam:beam-sdks-java-extensions-sql-zetasql:jar:2.20.0-SNAPSHOT (compile) > org.apache.beam:beam-sdks-java-extensions-sql:jar:2.20.0-SNAPSHOT (compile) > com.alibaba:fastjson:jar:1.2.49 (compile) > org.springframework.data:spring-data-redis:jar:1.8.6.RELEASE (provided?) > org.springframework.data:spring-data-keyvalue:jar:1.2.6.RELEASE (compile) > org.springframework.data:spring-data-commons:jar:1.13.6.RELEASE (compile) > org.springframework.hateoas:spring-hateoas:jar:0.23.0.RELEASE (compile?) > org.springframework:spring-aop:jar:4.3.5.RELEASE (compile) > com.jamonapi:jamon:jar:2.81 (compile?) > org.springframework:spring-context:jar:4.0.2.RELEASE (provided) > org.jruby:jruby:jar:1.7.10 (compile?) > org.jruby:jruby-stdlib:jar:1.7.10 (compile) > org.jruby:openssl:jar:0.9.4 (provided)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerArguments.getInputClasspath(LinkageCheckerArguments.java:209)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerArguments.getEntryPointJars(LinkageCheckerArguments.java:230)
        at com.google.cloud.tools.opensource.classpath.LinkageCheckerMain.main(LinkageCheckerMain.java:74)

> Task :checkJavaLinkage FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkJavaLinkage'.
> Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 6m 54s
906 actionable tasks: 622 executed, 284 up-to-date

```

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1242

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java [hunk 1]
```diff
@@ -60,7 +58,7 @@
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
   private final CommandLine commandLine;
-  private ImmutableList<Path> cachedInputClasspath;
+  private ClassPathResult classPathResult;
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java]: It feels very strange that there's a classpath result hiding inside the command line arguments class. Is there a more appropriate abstraction here?

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java]: I've been feeling that strangeness. Let me update them.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java [hunk 2]
```diff
@@ -190,47 +184,15 @@ private static Options configureOptions() {
...
-   */
-  ImmutableList<Path> getInputClasspath() throws RepositoryException {
-    if (cachedInputClasspath != null) {
-      // Avoid unnecessary dependency resolution, which Cloud OSS BOM takes around 4 seconds
-      return cachedInputClasspath;
-    }
-
-    if (commandLine.hasOption("b") || commandLine.hasOption("a")) {
-      List<Artifact> artifacts = getArtifacts();
-      ClassPathResult result = classPathBuilder.resolve(artifacts);
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java]: Now it's LinkageCheckerMain's responsibility to resolve the class path through dependency graphs.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java [hunk 3]
```diff
@@ -658,44 +663,32 @@ public void testGenerateInputClasspathFromLinkageCheckOption_mavenCoordinates_mi
...
-      Assert.fail(
-          "Because the unavailable dependency is not optional, it should throw an exception");
-    } catch (RepositoryException ex) {
-      Truth.assertThat(ex.getMessage())
-          .startsWith("Unresolved artifacts: org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4");
-    }
-  }
-
-  @Test
-  public void testGenerateInputClasspath_jarFileList()
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java]: This test has been moved to `LinkageCheckerArgumentTest.testGetJarFiles_jarFileList`

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java [hunk 4]
```diff
@@ -60,12 +56,10 @@
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
   private final CommandLine commandLine;
-  private ImmutableList<Path> cachedInputClasspath;
   private ImmutableList<Artifact> cachedArtifacts;
   private final ImmutableList<String> extraMavenRepositoryUrls;
   private final boolean addMavenCentral;
   private final boolean reportOnlyReachable;
-  private final ClassPathBuilder classPathBuilder = new ClassPathBuilder();
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java]: LinkageCheckerArguments is no longer responsible to resolve the class path.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java [hunk 5]
```diff
@@ -60,12 +56,10 @@
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
   private final CommandLine commandLine;
-  private ImmutableList<Path> cachedInputClasspath;
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerArguments.java]: LinkageCheckerArguments is no longer responsible to resolve the class path. Therefore it does not know class path (`cachedInputClasspath` variable).

**Pull request merged**